### PR TITLE
Add async templates

### PR DESCRIPTION
**Pull request by ViKomprenas**
This lets templates use `await` expressions, if ejs is given the option `async` and the environment supports it. Naturally, doing so changes the template function's return value to a `Promise<string>`.

### CODE REVIEW

**Comment by mde** [comment]
What is the advantage of async rendering? Other than filesystem access for includes (which might be a separate discussion), there should not be I/O happening as part of rendering a template. You should generally be getting your data up front, and passing it to the render function. Why would it be useful for the rendering process to be non-blocking?

**Comment by ViKomprenas** [comment]
I'm writing a CMS. The templates are supplied by the user and I can't know in advance what they want, beyond that it will involve accessing the contents of pages. While I can do that in advance for the page templates, I can't for listing templates, since they might want to filter the list arbitrarily and preloading *every* entry would be impractical.

**Comment by mde** [comment]
I must be missing something here. Async code-paths can invoke sync functions (but not the other way around), and you literally can't render an EJS template until you have supplied it with both the template text and the data. Both of those things may end up being acquired in an async fashion, but the template can't actually be rendered without them.

Unless there is something asynchronous happening inside of a function, I can't really see a benefit to making it an async function. And (aside from `include` calls that involve getting a template off disk), there's nothing in template rendering that needs to be asynchronous. EJS literally builds the source for an executable function out of the template and just runs it with the provided data.

Could you provide some sort of minimal example?

**Comment by ViKomprenas** [comment]
I'm passing the async functions in as part of the template parameters.

The minimal template I'm testing this with is

```
<% if (members) { -%>
    <ol>
        <% for (const member of members) { -%>
            <li><a href="/<%= member.name %>"><%= (await member.resolve()).title %></a></li>
        <% } -%>
    </ol>
<% } -%>
```

which should make the use case clear. `members`, here, is the pages to be displayed in the listing. It's an array of page _references_, not pages themselves, for the reasons I gave before.

**Comment by mde** [comment]
I think I understand better now. This looks like you're doing a bunch of I/O inside your template, which is generally considered bad practice -- the type of thing that has spawned the religiously "logic-less template" approach you see with Mustache et al.

Is there some specific reason you can't filter down your list of members, then do your individual member lookups in your data layer, then pass all that data down to the template for rendering? That would be the normal approach I would expect to see in a case like this. Clearly you are passing `members` down to EJS, so you have that filtered list already.

**Comment by RyanZim** [comment]
@mde On the server side, there could be a slight performance boost to doing it this way, because EJS can start working on parsing, etc. before all the data is fetched. Not sure how much perf boost this actually gives. Just throwing my 2 cents here.

Disclaimer: I haven't looked at the code in this PR yet, so not an endorsement of these changes.

**Comment by ViKomprenas** [comment]
Again, the template is supplied by the user, and `members` isn't filtered in any way. I left that up to the template deliberately, as the CMS itself is aiming to be pretty minimalist. Not even pagination is done beforehand, as that would impose a URL structure on the template. (Yes, yes, the database query would do it and be faster than anything JS-based. I'm not using a database.)

I *could* put a field in the template's header for code to run to filter `members`, but that wouldn't really be any different from doing it in the template directly, from a structural standpoint, and would require significantly more work than making this PR did.

**Comment by mde** [comment]
@RyanZim, despite any small perf gains, the question is specifically whether or not we open the door to async operations during the rendering phase. Async almost always implies some sort of I/O, and doing that sort of thing (DB queries, HTTP requests, etc.) inside your template is generally a bad practice.

OTOH, I've always been an advocate of letting people make their own decisions, and counting on people being grownup in their use of EJS, rather than putting artificial constraints in place to try to encourage good behavior. Does anybody else have an opinion on this?

**Comment by RyanZim** [comment]
> doing that sort of thing (DB queries, HTTP requests, etc.) inside your template is generally a bad practice

**Comment by mde** [comment]
@RyanZim What happens if the Promise rejects instead of settling? The idea of Promise/Deferred/Future is that you have a placeholder for a value, but it's not usually that straightforward in the real world, when errors occur. But having said that, this behavior is opt-in, and the code itself isn't super intrusive, so I'm inclined to merge this. Finally catching up on this stuff, so I'll try to get to it today.

File: lib/ejs.js [hunk 1]
```diff
@@ -581,7 +584,25 @@ Template.prototype = {
     }
 
     try {
-      fn = new Function(opts.localsName + ', escapeFn, include, rethrow', src);
+      var constructor;
```
[the remark this answers is not in the document]
Comment (ViKomprenas) [reply lib/ejs.js]: Right, right. I'm used to `let`. Fixed.

**Comment by ViKomprenas** [comment]
If you `await` a promise and it rejects, the promise you are in also rejects, which I just confirmed works as expected in templates.

**Comment by RyanZim** [comment]
@mde Yeah, if a value rejects, it should make the entire render fail.

**Comment by mde** [comment]
@ViKomprenas Thanks for this! Sorry for the delay in merging. I wanted to make one more change -- using a generated function instead of `eval`, hopefully to head off some problems with linters.

**Pull request merged**